compiler/circle-mpqsolver/src/Methods/BisectionSolver.cpp

```diff
@@ -0,0 +1,291 @@
+    if (arity == 0 && input_names.find(name) != input_names.end())
+    {
+      inputs.insert(node);
+    }
+    for (uint32_t inp_index = 0; inp_index < arity; ++inp_index)
+    {
+      auto arg = loco::must_cast<luci::CircleNode *>(node->arg(inp_index));
+      inputs_to_outputs[arg].push_back(node);
+    }
+  }
+
+  // initializing
+  std::vector<luci::CircleNode *> to_process{inputs.begin(), inputs.end()};
+  std::map<std::string, float> named_depth;
+  for (auto &node : inputs)
+  {
+    nodes_depth[node] = 0.f;
+    named_depth[node->name()] = 0.f;
+  }
+
+  // enumerating
+  while (!to_process.empty())
+  {
+    auto cur_node = to_process.back();
+    to_process.pop_back();
+    auto iter = nodes_depth.find(cur_node);
+    if (iter == nodes_depth.end())
+    {
+      return EXIT_FAILURE; // unexpected
+    }
+    float cur_depth = iter->second + 1;
+    // processing children
+    auto &children = inputs_to_outputs[cur_node];
+    for (auto &child : children)
+    {
+      auto node_depth = nodes_depth.find(child);
+      if (node_depth == nodes_depth.end() || node_depth->second < cur_depth)
+      {
+        // initialize depth
+        nodes_depth[child] = cur_depth;
+        to_process.push_back(child);
+        named_depth[child->name()] = cur_depth;
+      }
+    }
+  }
+
+  auto minmax = std::minmax_element(
+    nodes_depth.begin(), nodes_depth.end(),
+    [=](const std::pair<luci::CircleNode *, float> &el1,
+        const std::pair<luci::CircleNode *, float> &el2) { return el1.second < el2.second; });
+
+  min_depth = minmax.first->second;
+  max_depth = minmax.second->second;
+
+  return EXIT_SUCCESS;
+}
+
+bool error_at_input_is_larger_than_at_output(const NodeDepthType &nodes_depth, float cut_depth)
+{
+  float error_at_input = 0;
+  float error_at_output = 0;
+  for (auto &iter : nodes_depth)
+  {
+    float cur_error = ErrorApproximator::Approximate(iter.first);
+    if (iter.second < cut_depth)
+    {
+      error_at_input += cur_error;
+    }
+    else
+    {
+      error_at_output += cur_error;
+    }
+  }
+  std::cerr << error_at_input << " vs " << error_at_output << std::endl;
+  return error_at_input > error_at_output;
+}
+
+inline float ComputeLinearRoot(float target, float min_x, float max_x, float y_min_x, float y_max_x)
+{
+  float d_x_coeff = 1.f / (max_x - min_x);
+  float koeff = (y_max_x - y_min_x) * d_x_coeff;
+  float free_term = (max_x * y_min_x - min_x * y_max_x) * d_x_coeff;
+  float res = (target - free_term) / koeff;
+  return res;
+}
+
+} // namespace
+
+BisectionSolver::BisectionSolver(const std::string &input_data_path, float accuracy_ratio)
+  : _input_data_path(input_data_path), _accuracy_ratio(accuracy_ratio), _accuracy(0.f)
+{
+}
+
+int BisectionSolver::run(const std::string &input_model)
+{
+  // Load model from the file
+  luci::ImporterEx importerex;
+  auto module = importerex.importVerifyModule(input_model);
+  if (module.get() == nullptr)
+    return EXIT_FAILURE;
```

Comment:
(optional) It seems better to move the code that loads the model outside of this function for SRP.